aten/src/ATen/native/SharedReduceOps.h

```diff
@@ -122,20 +122,20 @@ struct WelfordOps {
   }
 };
 
-template <typename acc_t, typename factor_t>
+template <typename acc_t, typename data_t, typename out_t, typename factor_t>
 struct MeanOps {
   factor_t factor;
 
-  inline C10_DEVICE acc_t reduce(acc_t a, acc_t b, int64_t /*idx*/) const {
-    return combine(a, b);
+  inline C10_DEVICE acc_t reduce(acc_t a, data_t b, int64_t /*idx*/) const {
```

Comment:
I don't like the approach of adding a lot of template parameters here. Note, MeanOps and NormOps today are perfectly capable of handling fp32 accumulation with low precision data type, they've been doing it for a long time. This is achieved via implicit conversion of the `scalar_t` values read from the tensor to `acc_t` and instantiating corresponding reduction templates (`MeanOps`, `NormOps`, reduction operations going through function wrapper with generic definitions of reduce, combine and project) with the proper `acc_t`. The accuracy is never lost, because conversion happens as soon as is feasible, and `combine` and `reduce` both accept only `acc_t` arguments for this reason. If BFloat cannot be implicitly converted into `acc_t` which is float, that is a problem with `BFloat` implementation and should be fixed there, not in the reduction functions.  Same for `project`, there is no reason to return `out_t` if that conversion can happen implicitly later.

Reply:
Sincerely thanks for your comments. It is a design choice here. It follows the design which described at [Reduce.h](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cpu/Reduce.h#L154)

```
ops_t is such that &ops_t::reduce, &ops_t::combine, and &ops_t::project exist and satisfy the following.
reduce: (acc_t, data_t, index_t) -> acc_t adds one data point to the accumulated value.
combine: (acc_t, acc_t) -> acc_t combines two accumulated values into one.
project: acc_t -> out_t finishes the reduction, getting the required output.
```

The design above may be more flexible, more readable and more straightforward than "implicit conversion".

Reply:
The design above introduces flexibility where it is not needed, and is arguably less readable and less straightforward (note, MeanOps and NormOps were already doing what you want, you did not need to change them).

Reply:
Current implementation is not robust. It may bring memory corruption if introduces accumulate type.

The [caller](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cpu/ReduceOpsKernel.cpp) only passes `scalar_t` to the reduce kernel and then call `binay_kernel_reduce` to extecute it. For [example](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cpu/ReduceOpsKernel.cpp#L116)
```C++
static void mean_kernel_impl(TensorIterator& iter) {
  AT_DISPATCH_ALL_TYPES_AND_COMPLEX(iter.dtype(), "mean_cpu", [&] {
    scalar_t factor = scalar_t(iter.num_output_elements()) / scalar_t(iter.numel());
    binary_kernel_reduce(
      iter,
      MeanOps<scalar_t, scalar_t> {factor},
      scalar_t(0)
    );
  });
}
```
For low precision type, the reduce kernel needs to use accumulate type(`acc_t`) of data type(`data_t`) for computation to keep precision and then project back to output type(`out_t`). 
And the framework does not depends on "implicit conversion" to convert `acct_t` to `out_t`. It call  [`set_results`](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cpu/Reduce.h#L121) to convert and store the projected result.
`set_results<r_traits>(ops.project(total_acc), sub_iter, num_outputs);`

The `set_result` definitions are as the follows.
```C++
template<typename traits, typename res_t>
static void set_result(const int index, const res_t result, const TensorIterator &iter, const int num_outputs) {
  // static_assert(std::is_same<res_t, typename traits::arg2_t>::value, "data types must match");
  if (index < num_outputs) {
    char *out = (char *) iter.data_ptr(index);
    *(res_t *) out = result;
  }
}

template<typename traits, typename res_t>
static void set_results(const res_t result, const TensorIterator &iter, const int num_outputs) {
  AT_ASSERT(num_outputs == 1);
  set_result<traits>(0, result, iter, num_outputs);
}
```
The template parameter `res_t` will be set to the result type of `project` by "Automatic Template Argument Deduction". There is a potential issue here because most of the reduce kernels only accept 1 template parameter. It means that the result type of `project` will be `acc_t` if the caller uses `acc_t` for computation. Take `BFloat16` as the example, its accumulate type  is `float` and then the core code fragment of the above code
```C++
char *out = (char *) iter.data_ptr(index);
*(res_t *) out = result;
```
will be initialized as
```C++
char *out = (char *) iter.data_ptr(index);
*(float *) out = result;
```
But the data type of `out` is BFloat16. It will cause memory corruption.

Besides that, I think the design should be consistent. [CUDA](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/ReduceOpsKernel.cu) also follows the design.

Reply:
I agree that design should be consistent, but cuda does not follow the same design. You had to change all the template arguments in ReduceOpsKernel.cu, but cuda reduction kernels already supported all the functionality that you seek here, so this change should not have been necessary.

Reply:
Sorry, I may not capture your point.
```C++
template <typename scalar_t, typename acc_t=scalar_t, typename out_t=scalar_t>
void mean_kernel_impl(TensorIterator& iter) {
  float factor = float(iter.num_output_elements()) / iter.numel();
  gpu_reduce_kernel<scalar_t, out_t>(iter, MeanOps<acc_t, float> {factor});
}
```
And the caller code is
```C++
static void mean_kernel_cuda(TensorIterator& iter) {
  if (iter.dtype() == kHalf) {
    return mean_kernel_impl<at::Half, float>(iter);
  } else if (iter.dtype(1) == kHalf && iter.dtype() == kFloat) {
    // type promotion that does cast and reduction in a single kernel
    return mean_kernel_impl<at::Half, float, float>(iter);
  }
  AT_DISPATCH_ALL_TYPES(iter.dtype(), "mean_cuda", [&]() {
    mean_kernel_impl<scalar_t>(iter);
  });
}
```
The CUDA code also clearly defines the template parameter `scalar_t`(`data_t`), `acc_t` and `out_t` here and follows the design of [Reduce.h](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cpu/Reduce.h#L154)
```
ops_t is such that &ops_t::reduce, &ops_t::combine, and &ops_t::project exist and satisfy the following.
reduce: (acc_t, data_t, index_t) -> acc_t adds one data point to the accumulated value.
combine: (acc_t, acc_t) -> acc_t combines two accumulated values into one.
project: acc_t -> out_t finishes the reduction, getting the required output
```

Reply:
Current gpu code calls MeanOps and the like with 2 template parameters (`<acc_t, float>`) rather than 4 as would be the case after this PR.

Reply:
The second template parameter of `MeanOps` is `factor_t` and it is not under our discussion. Let's focus on the first template parameter of `MeanOps`.

Current cuda code does not only depend on the template parameters of `MeanOps` to get `acc_t`, but also it passes `scalar_t`(`data_t`) and `out_t` to `gpu_reduce_kenerl`. Then the `acc_t`, `data_t`, `out_t` is clearly defined here and its source is the template parameters of `mean_kernel_impl`. (We can refer to the implementation of [`gpu_reduce_kernel`](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/Reduce.cuh#L671) for more details.)
```C++
gpu_reduce_kernel<scalar_t, out_t>(iter, MeanOps<acc_t, float> {factor});
```
```C++
template <typename scalar_t, typename out_scalar_t, int vt0=4, typename ops_t, typename ident_t=double>
inline void gpu_reduce_kernel(TensorIterator& iter, const ops_t& ops, ident_t ident=0,
                              AccumulationBuffer* acc_buf_ptr=nullptr, int64_t base_idx=0)
```
```C++
template <typename scalar_t, typename acc_t=scalar_t, typename out_t=scalar_t>
void mean_kernel_impl(TensorIterator& iter)
```

But for CPU path, there is no place to clearly define `acc_t`, `data_t` and `out_t`. So we follow the design to clearly define `acc_t`, `data_t` and `out_t`.

Reply:
Current CUDA code also defined `data_t`, `acc_t` and `out_t`, but the three data types split into  `gpu_reduce_kernel` and `MeanOps`.